### PR TITLE
bpo-27752: improve documentation of csv.Dialect

### PR DESCRIPTION
**Pull request by jdevries3133**
Note that **all content changes** are in the first commit. The second commit is 100% formatting of the `rst` file. There are also some very minor proofreading corrections in the final commit.

I'm very open to feedback on this; please let me know if there are any bits in here that you feel could be communicated better!

https://bugs.python.org/issue27752

### CODE REVIEW

**Comment by iritkatriel** [comment]
> Note that **all content changes** are in the first commit. The second commit is 100% formatting of the `rst` file. There are also some very minor proofreading corrections in the final commit.

The formatting (whitespace?) changes are unlikely to be approved. You should limit this PR to the content change you are making (possibly fix typos in other places).

The PR will be squashed into one commit, so the diff will be huge and hide the actual change.  Sweeping whitespace-type changes are not going to be accepted in a separate PR either because they spam the commit log and the "blame" outputs.

File: Doc/library/csv.rst [hunk 1]
```diff
@@ -225,15 +225,30 @@ The :mod:`csv` module defines the following classes:
...
 .. class:: Dialect
 
-   The :class:`Dialect` class is a container class relied on primarily for its
-   attributes, which are used to define the parameters for a specific
-   :class:`reader` or :class:`writer` instance.
+   The :class:`Dialect` class is a container class whose attributes contain
+   information for how to handle doublequotes, whitespace, delimiters, etc.
+   Due to the lack of a strict csv specification, different applications
+   produce subtly different csv data.  :class:`Dialect` attributes
+   are used to define the parameters for a specific :class:`reader` or
```
Comment (jdevries3133) [inline Doc/library/csv.rst]: Should say ... Determine how reader or writer instances behave

File: Doc/library/csv.rst [hunk 2]
```diff
@@ -65,8 +65,8 @@ The :mod:`csv` module defines the following functions:
    the :class:`Dialect` class or one of the strings returned by the
    :func:`list_dialects` function.  The other optional *fmtparams* keyword arguments
    can be given to override individual formatting parameters in the current
-   dialect.  For full details about the dialect and formatting parameters, see
-   section :ref:`csv-fmt-params`.
+   dialect.  For full details about dialects and formatting parameters, see
```
Comment (jdevries3133) [inline Doc/library/csv.rst]: Still quite a few whitespace issues to fix

File: Doc/library/csv.rst [hunk 3]
```diff
@@ -253,9 +268,8 @@ The :mod:`csv` module defines the following classes:
 
 .. class:: Sniffer()
 
-   The :class:`Sniffer` class is used to deduce the format of a CSV file.
-
-   The :class:`Sniffer` class provides two methods:
+   The :class:`Sniffer` class is used to deduce the format of a CSV file.  It
```
Comment (jdevries3133) [inline Doc/library/csv.rst]: Maybe should say 'the structure' instead. There is no formal csv file format

File: Doc/library/csv.rst [hunk 4]
```diff
@@ -318,18 +332,22 @@ The :mod:`csv` module defines the following exception:
 
 .. _csv-fmt-params:
 
-Dialects and Formatting Parameters
+Custom Dialects
 ----------------------------------
 
-To make it easier to specify the format of input and output records, specific
-formatting parameters are grouped together into dialects.  A dialect is a
-subclass of the :class:`Dialect` class having a set of specific methods and a
```
Comment (jdevries3133) [inline Doc/library/csv.rst]: Looking at this now, the original docs look better than my former proposed changes in this spot... I will revise this

**Review by jdevries3133 — COMMENTED** [review]
Don't mind me; just talking to myself. When this was marked stale,I opened it up and saw these issues. I will fix them soon, but don't want anyone to waste their time reviewing in the interim.

**Comment by iritkatriel** [comment]
> Don't mind me; just talking to myself. When this was marked stale,I opened it up and saw these issues. I will fix them soon, but don't want anyone to waste their time reviewing in the interim.

You can mark the PR as draft and then people won't review it until you say it's ready.

**Comment by jdevries3133** [comment]
> You can mark the PR as draft and then people won't review it until you say it's ready.

@iritkatriel Yes, well the other factor here is that I was reading the diff just before bed and wanted to leave myself the notes as breadcrumbs, too!

Anyway, this PR is ready for another review, I think it is ready to go. The diff is much smaller as I got rid of *all* whitespace changes, plus a lot of pedantic changes that I can now see were unnecessary. A summary of the changes now are:

- fix mechanics in two places: change "the dialect" to "dialects" when speaking about dialect classes, in general
- revise and expand the paragraph of documentation for the `Dialect` class to make it more descriptive and less technical-jargon dense
- create hyperlink by changing change "dialect" to "class:Dialect" where appropriate

File: Doc/library/csv.rst [hunk 5]
```diff
@@ -225,9 +225,21 @@ The :mod:`csv` module defines the following classes:
 
 .. class:: Dialect
 
-   The :class:`Dialect` class is a container class relied on primarily for its
-   attributes, which are used to define the parameters for a specific
-   :class:`reader` or :class:`writer` instance.
+   The :class:`Dialect` class is a container class whose attributes contain
+   information for how to handle doublequotes, whitespace, delimiters, etc.
+   Due to the lack of a strict csv specification, different applications
```
Comment (ambv) [inline Doc/library/csv.rst]:
```suggestion
   Due to the lack of a strict CSV specification, different applications
```

File: Doc/library/csv.rst [hunk 6]
```diff
@@ -225,9 +225,21 @@ The :mod:`csv` module defines the following classes:
...
+   produce subtly different csv data.  :class:`Dialect` instances define how
```
Comment (ambv) [inline Doc/library/csv.rst]:
```suggestion
   produce subtly different CSV data.  :class:`Dialect` instances define how
```

**Review by ambv — APPROVED** [review]

**Pull request merged**